Allow readers to take a custom Record class

billdueber opened the pull request:
Most uses of a MARC record require extraction of things like OCLC numbers, local identifiers, etc. The ruby-marc reader classes hard-code the use of MARC::Record, which means folks who want to do that either need to inline their code, wrap a MARC::Record object, or call
the reader.

This code allows the binary reader and the non-REXML XML readers to take a `record_class` option. The
required interface is for the passed class is:

* `#append_control_field(tag, value)`
* `#append_data_field(tag, ind1, ind2, subfields)` where `subfields` is an array of [code, value] pairs
* `#leader` and `#leader=`
* `#new_from_hash(marc_hash)` which is used in the XML reader to avoid using specific field classes

All the rest of the code has been modified to use just those methods when building up a record.

The easiest way to take advantage of this would be to just subclass MARC::Record, e.g.

```
class MyRecord < MARC::Record
  def isbns
    m = /[\d\-Xx]{10,15}/
    fields('020').map{|f| f['a'].scan(m).first.gsub('-', '').upcase}
  end
end

r = MARC::Reader.new("test/batch.dat", record_class: MyRecord).first
r.isbns #=> ["0471383147"]

```

dmolesUC commented:
I like the idea of making `Record` more abstract and making it easier to provide alternate implementations, but hard-coding a record class strikes me as a little clunky. Maybe it's my Java background, but I'd rather see an optional ~~factory method~~ block passed in that could return whatever class it wanted to, not necessarily the same one every time.

Also, if I understand correctly (and I may not, this is just a quick glance), `new_from_hash` is only used in `JSONLReader`, and `JSONLReader` always uses it (and never the regular initializer), which even more suggests passing in a block.

billdueber commented:
Right -- `new_from_hash` is the (opaquely-named) code that deals with the
marc-in-json format, so as you'd expect, that's the only place it's used.
To use it "by hand" you would use `record =
MARC::Record.new_from_hash(JSON.parse(marc_in_json_string))`. Oh, I guess I
added

I _kind of_ sympathize with the factory object, but all it would be is a
lambda or something that guarantees a particular interface for creating
records, and that's basically what I have now.  Can you walk me through why
a factory object would be more useful than a class with a documented API?
Without the formal interfaces and static typing in a language like Java,
I'm not sure I see the point, but I remain convincible.

(Side note: I _swear_ I thought I'd put in `Record.new_from_json(str)` in
there during the last go-round, but apparently not.  I might sneak that in
with the, too. `new_from_marc_in_json` isn't exactly concise, but it's
certainly more easily understood than `new_from_hash`)

On Thu, May 16, 2024 at 1:44 PM David Moles ***@***.***>
wrote:

> I like the idea of making Record more abstract and making it easier to
> provide alternate implementations, but hard-coding a record class strikes
> me as a little clunky. Maybe it's my Java background, but I'd rather see an
> optional factory method block passed in that could return whatever class
> it wanted to, not necessarily the same one every time.
>
> Also, if I understand correctly (and I may not, this is just a quick
> glance), new_from_hash is only used in JSONLReader, and JSONLReader
> always uses it (and never the regular initializer), which even more
> suggests passing in a block.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ruby-marc/ruby-marc/pull/107#issuecomment-2115849161>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA32VQSOP7PWNWZK2YGU2LZCTV7NAVCNFSM6AAAAABHYUX74GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMJVHA2DSMJWGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Bill Dueber
Library Systems Programmer
University of Michigan Library

billdueber commented:
Oh, and I missed what you were probably _really_ pointing out, which is that `Record#new_from_hash` is also a requirement for any `MARC::Record` overrides because of `jsonl_reader`.  And `new_from_marchash` which is used as an intermediate format in the XML readers.  And `Record#to_hash`, which produces the marc-in-json structure for the `jsonl_writer`. 

It would be easier to say, "It's gotta be a subclass of MARC::Record" and be done with it. A truly alternate implementation would have to override everything anyway, and the subclass requirement implies all the methods must work as-is or be overridden.

dmolesUC commented:
Oh, I missed `new_from_marchash`, didn't I? (Git blame says it's been there forever, but isn't currently called on main except from tests?)

Yeah, if `new_from_hash` and `new_from_marchash` weren't umpteen years old already, I'd say these are signs they don't belong in `MARC::Record` at all, and should just be part of `JSONLReader` / `GenericPullParser` (respectively). And unfortunately, saying "it needs to subclass `MARC::Record`" doesn't really help much, b/c if you don't re-declare those methods in your own subclass, when the reader calls `@record_class.new_from_hash` or `@record_class.new_from_marchash`, it'll end up just getting the original `MARC::Record`. (I'm sure there's some eldritch metaprogramming way around that, but it's probably not a good idea.)

One possible refactoring, assuming we keep the `@record_class` idea, would be something like:

1. move the meat of `new_from_hash` into a class method on `JSONLReader`, and the meat of `new_from_marchash` into a class method on `GenericPullParser`, something like `populate_from_hash(marc_record, h)`
2. change `new_from_hash` and `new_from_marchash` to just be simple wrappers around those (`JSONLReader.populate_from_hash(self.new, h)` / `GenericPullParser.populate_from_hash(self.new, mh)`)
3. change `JSONLReader` and `GenericPullParser` to call the `populate` methods, passing `@record_class.new`
4. deprecate `new_from_hash` and `new_from_marchash`

That gets us around the requirement of the alternative implementation having to not only implement a (duck-typed) interface, but also override a bunch of factory methods.

(If there are any other ways we instantiate a `MARC::Record` that I'm missing, the same approach should work there.)

> Can you walk me through why a factory object would be more useful than a class with a documented API?

The main advantage of a factory object / factory lambda / factory block / whatever is that it doesn't always have to return the same type -- not that I have a specific example in mind of why you'd want to do something different, particularly while reading the same input document; and for everything *but* `#new_from_hash` and `#new_from_marchash`, you wouldn't actually have any information available to decide what type you wanted to instantiate anyway. But in principle, and maybe more so in languages other than Ruby, I can imagine you have some specialized record class that you only want to use sometimes, for certain types of records. (IIRC this is actually the original justification from the Gang of Four or whoever for static factory methods of the `#new_from_hash` type, too, that under the hood they'd return different implementations. Not really relevant to the user-extensible implementation situation, though.)

FWIW, and I'm not saying this is a great solution, but toward the original problem, what I ended up doing at Berkeley was just [monkey-patching extra methods into `MARC::Record`](https://github.com/BerkeleyLibrary/marc/blob/0.3.1/lib/marc_extensions/record.rb) (and then in some cases [monkey-patching `MARC::XMLReader` to call those methods](https://github.com/BerkeleyLibrary/marc/blob/0.3.1/lib/marc_extensions/xml_reader.rb).

billdueber commented:
> Oh, I missed `new_from_marchash`, didn't I? (Git blame says it's been there forever, but isn't currently called on main except from tests?)

Right. `new_from_hash` and `new_from_marchash` are the primary ways people deal with marc-in-json and marcjson respectively. They pre-date the json-oriented `Reader`s and can't be moved at this point. 

>  And unfortunately, saying "it needs to subclass `MARC::Record`" doesn't really help much, b/c if you don't re-declare those methods in your own subclass, when the reader calls `@record_class.new_from_hash` or `@record_class.new_from_marchash`, it'll end up just getting the original `MARC::Record`

No, I think that's what you're missing. Any call to `@record_class.new_from_*` is going to give you an instance of `@record_class`. That's what most of this pull does -- track down all the ways various spots were calling `MARC::Record` or `MARC::DataField` explicitly and channeling it all through the passed `@record_class`. Calling `MyRecordClass.new_from_hash` is going to give you a `MyRecordClass` object, and passing `MyRecordClass` to any of the readers is going to give you a stream of `MyRecordClass` objects. The only way to get raw `MARC::Record` objects is to use it directly from your own code, or pass it (implicitly or explicitly) to a reader. 

> (other stuff)
> That gets us around the requirement of the alternative implementation having to not only implement a (duck-typed) interface, but also override a bunch of factory methods.

I guess if your alternate implementation is totally from scratch, then yeah. If it's just a subclass, you won't have to override anything unless you wants a different implementation of the functionality (e.g., changing how fields are indexed internally or something). 

The idea of wanting to return different record implementations _within a single reader_ is kind of theoretically intriguing, but I'm having trouble finding a use case. I mean, you'd have to parse the raw data enough to figure out what kind of thing you want to instantiate and then create it. I guess maybe a file that mixes bib and holdings records? It's not an issue that seems strong enough to rewrite everything to accommodate it. 

I've passed factories around (yay, unto the FactoryFactoryFactory) in java code when necessary, but I think it solves a problem that we just don't have in duck-typed languages. 

> FWIW, and I'm not saying this is a great solution, but toward the original problem, what I ended up doing at Berkeley was just [monkey-patching extra methods into `MARC::Record`](https://github.com/BerkeleyLibrary/marc/blob/0.3.1/lib/marc_extensions/record.rb) (and then in some cases [monkey-patching `MARC::XMLReader` to call those methods](https://github.com/BerkeleyLibrary/marc/blob/0.3.1/lib/marc_extensions/xml_reader.rb)).

Well, now I just want to punch you. Because I have never, ever....errrrr...ever....ummmm...never mind.

dmolesUC commented:
> Well, now I just want to punch you. Because I have never, ever....errrrr...ever....ummmm...never mind.

Fair. :)

> Any call to `@record_class.new_from_*` is going to give you an instance of `@record_class`.

Without explicitly re-implementing it? When I try adding this to `tc_hash.rb` (on the `record_as_factory` branch):

```ruby
  def test_subclass
    my_record_subclass = Class.new(MARC::Record)
    raw = IO.read("test/one.dat")
    r = my_record_subclass.new_from_marc(raw)
    assert_kind_of(my_record_subclass, r)
  end
```

it fails with:

```none
was expected to be kind_of?
  <#<Class:0x0000000106e40dc0>> but was
  <MARC::Record>.
/Users/david/Work/ruby-marc/test/tc_hash.rb:28:in `test_subclass'
     25:     my_record_subclass = Class.new(MARC::Record)
     26:     raw = IO.read("test/one.dat")
     27:     r = my_record_subclass.new_from_marc(raw)
  => 28:     assert_kind_of(my_record_subclass, r)
     29:   end
     30:     
     31: end
```

The same thing happens if I explicitly create a named class with `class MyRecordSubclass < MARC::Record; end` (but better not to clutter up the global namespace):

```none
was expected to be kind_of?
  <TestHash::MyRecordSubclass> but was
  <MARC::Record>.
/Users/david/Work/ruby-marc/test/tc_hash.rb:30:in `test_subclass'
     27:   def test_subclass
     28:     raw = IO.read("test/one.dat")
     29:     r = MyRecordSubclass.new_from_marc(raw)
  => 30:     assert_kind_of(MyRecordSubclass, r)
     31:   end
```

I agree passing a factory object / lambda / block / whatever into the reader initializer isn't particularly Ruby-ish, and with the current reader API passing in a record class might be the best solution, but only if we can get everything to go through `@record_class.new` (unless I'm wrong somehow about the above behavior).

adjam commented:
I think it might help to extract different use cases for why you'd want to customize record creation -- one is to produce objects with convenience methods, such as `#isbns` example, and another -- maybe a bit more speculative -- is to customize the creation process itself (e.g. if I'm going for minimalism maybe I just want the 001 and the 245, so I can build a ginormous list of records without using too much memory, I'unno.

It seems to me that the "factory" approach is most suited to the latter kind of thing; the "supply a class" should handle the former. But if the use cases for the 'customize the creation process' approach are, let's say _very_ speculative, then maybe just supplying a `MARC::Record` class at reader construction time is enough. 

The thing is, the only example I've cited is trivially achievable by overriding `#new_from_hash` and discarding all but the `001` and the `245`; although I guess if I put on my SERIOUSLY SPECULATIVE hat, I can imagine someone heck-bent on 'efficiency' wanting to elide all the other fields and just jump ahead to the end of the record once they've got what they need.

You could maybe get access to both -- while losing some coherence in the constructor signature -- by assuming a default class name and assuming a default factory that's just going to build instances of `record_class` in the same way.  

So someone who just wants to add `#isbns`-style conveniences can just supply their subclass at construction time, while someone who wants to break new grounds in record construction can supply (say) a block that does whatever it's gonna do.

adjam commented:
To make that last proposal slightly more concrete:

```ruby
MARC::Reader.new('some-marc.dat', record_class: MyRecordClass)
```

vs.

```ruby
MARC::Reader.new('some-marc.dat') do |thing|

dmolesUC commented:
(resisting the urge to start writing a streaming MARC parser)

billdueber commented:
I had adjusted everything to use the new class, but missed _passing_ that new class to `MARC::Reader.decode` from `#new_from_marc`. Fixed now. Will add tests after lunch.

billdueber commented:
For the "only the 245 and 001" case (which, let's be honest, might not be the most probable use case) you could also override `.create_data_field` and `.create_control_field` to ignore any tags you don't want. Which is also not a super-clean solution, but is at least easier than overriding all of `new_from_hash`. 

I'm not _against_ a factory object. I mean, ok, I kind of am, but mostly I'm afraid of the many bits that would have to be valid calls against that factory object.  

A meet-in-the-middle object would be to have MARC::Record comply with whatever the factory spec is, so you can do either.

I'm having trouble picturing how the factory object would work. Could someone throw out some pseudocode?

billdueber commented:
> (ruby-ish or not, I think I am somewhat attracted by the idea of "a reader" being flexible enough to produce instances of whatever it is I want from my MARC data. In part, the motivation here is that the underlying assumption that "a `MARC::Reader` produces `MARC::Record`s" seems to make more sense in a statically-typed context)

One of my concerns is that the marc-binary reader builds up the record on the fly, as did the XML pull-parser. This pull request changes the latter to build a marc_hash structure and then convert that at the end, but only because it was a little more convenient for and XML parsing is so slow that conversion time isn't a big thing.

I like the idea of the readers producing a common intermediate format -- I like marc_hash because it's dead-simple, rather that marc-in-json which requires a bunch of screwing around with one-pair-hashes for subfields -- but it may not make a difference performance-wise and the JSONL readers will be pulling in marc-in-json to begin with and can basically just pass it through.

So, that would mean
  * Every reader always builds a record internally as a `marc-in-json` hash
  * It passes it to `record_class#new_from_hash` (or anything that responds to `#new_from_hash`)

So now we're down to one method for a `Record`-alike (`#new_from_hash(marc-in-json-hash)`) and three for the reader interface:

```ruby
# Set up the reader. if a block is given, the new object will do whatever
# setup needs to be done, yield itself, and then close itself.
# @param inputfile [String,IO] An open IO object or a string naming a file on disk
# @param options [Hash] hash of options
# @option record_factory [#new_from_hash] An object that takes a hash and returns a record
# @yield [self.class] The reader object, so you can pass a block in
# @return [nil] Nothing? I'm not sure we're consistent in what we return now
def initialize(inputfile, options = {}, &blk)
  @factory = options[:record_factory] || MARC::Record
  # do the setup
  yield self if block_given?
  self.close
end

# Provide an iterator for objects returned by the passed record_factory
# For the stock implementation, this will be a MARC::Record object,
# but could be basically anything
# @yield [#new_from_hash] An object returned from record_factory.new_from_hash
# @return [nil]
def each
  return enum_for(:each) unless block_given?
  however_we_build_each_hash.each do
    yield @factory.new_from_hash(record_hash)
  end
  nil
end

# Close the reader
# @return [nil]
def close
  # do whatever's neccessary
  nil
end
```
I think that's it? 

Of course, we need to benchmark it for binary and XML.  And change the non-nokogirl XML parser implementations to throw an error if you pass them a `record_factory`, 'cause there's no way I'm wading into that mess.

billdueber commented:
OK, I did a little [benchmark](https://gist.github.com/billdueber/94b4a1315473411799a584c43f7ed1b2) that tries to see what would happen if we changed the binary reader to first target a marc-in-json hash, and then turn it into a record right before yielding.

The attempted comparison is (a) building a record as the parse happens, vs (b) building a marc-in-json hash and then converting. The latter means that the reader is building two objects: the initial hash, and then the record.

The very worst case scenario is that building up the marc-in-json hash instead of building the record _takes just as long_, then adding the overhead of decoding the marc-in-json. That seems ridiculous to me, but it's easy to benchmark.

You can see the whole thing at the gist above, but the top-line numbers, comparing them and doing a tiny bit of work in the loop to try and get a sense of "would this be a difference-that-makes-a-difference" in a real world use.

Top line numbers are this:

```text
Calculating -------------------------------------
           Just read      6.947k (± 9.8%) i/s -    137.856k in  20.031108s
       With a decode      5.867k (± 8.4%) i/s -    116.772k in  20.039091s

Comparison:
           Just read:     6946.7 i/s
       With a decode:     5867.0 i/s - same-ish: difference falls within error

```

So,  it almost certainly wouldn't slow things down enough to notice.

dmolesUC commented:
> > (resisting the urge to start writing a streaming MARC parser)
> 
> @dmolesUC I'm pretty sure all the readers _do_ stream. Or I'm misunderstanding you.

I don't actually remember for sure now what was going through my head when I wrote that, but I suspect I meant some kind of SAX-style event-based push parser.
